src/Storages/StorageDelta.cpp

```diff
@@ -0,0 +1,289 @@
+        if (!outcome.IsSuccess())
+            throw Exception(
+                ErrorCodes::S3_ERROR,
+                "Could not list objects in bucket {} with key {}, S3 exception: {}, message: {}",
+                quoteString(bucket),
+                quoteString(table_path),
+                backQuote(outcome.GetError().GetExceptionName()),
+                quoteString(outcome.GetError().GetMessage()));
+
+        const auto & result_batch = outcome.GetResult().GetContents();
+        for (const auto & obj : result_batch)
+        {
+            const auto & filename = obj.GetKey();
+
+            if (std::filesystem::path(filename).extension() == ".json")
+                keys.push_back(filename);
+        }
+
+        request.SetContinuationToken(outcome.GetResult().GetNextContinuationToken());
+        is_finished = !outcome.GetResult().GetIsTruncated();
+    }
+
+    return keys;
+}
+
+std::shared_ptr<ReadBuffer> JsonMetadataGetter::createS3ReadBuffer(const String & key)
+{
+    // TBD: add parallel downloads
+    return std::make_shared<ReadBufferFromS3>(
+        base_configuration.client,
+        base_configuration.uri.bucket,
+        key,
+        base_configuration.uri.version_id,
+        /* max single read retries */ 10,
+        ReadSettings{});
+}
+
+void JsonMetadataGetter::handleJSON(const JSON & json)
+{
+    if (json.has("add"))
+    {
+        auto path = json["add"]["path"].getString();
+        auto timestamp = json["add"]["modificationTime"].getInt();
+
+        metadata.setLastModifiedTime(path, timestamp);
+    }
+    else if (json.has("remove"))
+    {
+        auto path = json["remove"]["path"].getString();
+        auto timestamp = json["remove"]["deletionTimestamp"].getInt();
+
+        metadata.remove(path, timestamp);
+    }
+}
+
+StorageDelta::StorageDelta(
+    const S3::URI & uri_,
+    const String & access_key_,
+    const String & secret_access_key_,
+    const StorageID & table_id_,
+    const String & format_name_,
+    ColumnsDescription columns_,
+    const ConstraintsDescription & constraints_,
+    const String & comment,
+    ContextPtr context_)
+    : IStorage(table_id_)
+    , base_configuration({uri_, access_key_, secret_access_key_, {}, {}, {}})
+    , log(&Poco::Logger::get("StorageDeltaLake (" + table_id_.table_name + ")"))
+    , table_path(uri_.key)
+{
+    StorageInMemoryMetadata storage_metadata;
+    StorageS3::updateS3Configuration(context_, base_configuration);
+
+    JsonMetadataGetter getter{base_configuration, table_path};
+
+    auto keys = getter.getFiles();
+
+    auto new_uri = base_configuration.uri.uri.toString() + generateQueryFromKeys(std::move(keys));
+
+    LOG_DEBUG(log, "New uri: {}", new_uri);
+    LOG_DEBUG(log, "Table path: {}", table_path);
+    auto s3_uri = S3::URI(Poco::URI(new_uri));
+
+    if (columns_.empty())
+    {
+        columns_ = StorageS3::getTableStructureFromData(format_name_, s3_uri, access_key_, secret_access_key_, "", false, {}, context_);
```

Comment:
For table engine columns are always non empty. There might be no column for table function, but you did not currently implement table function, btw do you want to implement? it is very easy, as table function wrapped storageDeltaLake.  See `src/TableFunction/TableFunctionS3.cpp`

Reply:
In case of StorageDelta and StorageHudi columns may be empty because they support schema inference from data.
Will implement table functions for DeltaLake and Hudi engines a little bit later :)